galley/pkg/config/analysis/analyzers/virtualservice/jwtclaimroute.go

```diff
@@ -0,0 +1,142 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualservice
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/api/security/v1beta1"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	route2 "istio.io/istio/pilot/pkg/networking/core/v1alpha3/route"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+type JWTClaimRouteAnalyzer struct{}
+
+var _ analysis.Analyzer = &JWTClaimRouteAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *JWTClaimRouteAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "virtualservice.JWTClaimRouteAnalyzer",
+		Description: "Checks the VirtualService using JWT claim based routing has corresponding RequestAuthentication",
+		Inputs: collection.Names{
+			collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
+			collections.IstioSecurityV1Beta1Requestauthentications.Name(),
+			collections.IstioNetworkingV1Alpha3Gateways.Name(),
+			collections.K8SCoreV1Pods.Name(),
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *JWTClaimRouteAnalyzer) Analyze(c analysis.Context) {
+	c.ForEach(collections.IstioNetworkingV1Alpha3Virtualservices.Name(), func(r *resource.Instance) bool {
+		s.analyze(r, c)
+		return true
+	})
+}
+
+func (s *JWTClaimRouteAnalyzer) analyze(r *resource.Instance, c analysis.Context) {
+	// Check if the virtual service is using JWT claim based routing.
+	vs := r.Message.(*v1alpha3.VirtualService)
+	var vsRouteKey string
+	if vsRouteKey = routeBasedOnJWTClaimKey(vs); vsRouteKey == "" {
+		return
+	}
+	vsNs := r.Metadata.FullName.Namespace
+
+	// Check if the virtual service is applied to gateway.
+	var foundGateway bool
+	for _, gwName := range vs.Gateways {
+		if gwName == util.MeshGateway {
+			continue
+		}
+		foundGateway = true
+
+		gwFullName := resource.NewShortOrFullName(vsNs, gwName)
+		gwRes := c.Find(collections.IstioNetworkingV1Alpha3Gateways.Name(), gwFullName)
+		if gwRes == nil {
+			// The gateway does not exist, this should already be covered by the gateway analyzer.
+			continue
+		}
+
+		gw := gwRes.Message.(*v1alpha3.Gateway)
+		gwSelector := k8s_labels.SelectorFromSet(gw.Selector)
+
+		// Check each pod selected by the gateway.
+		c.ForEach(collections.K8SCoreV1Pods.Name(), func(rPod *resource.Instance) bool {
+			pod := rPod.Message.(*v1.Pod)
```

Comment:
The ForEach function allows for the possibility of iterators, which will not load the entire collection contents into memory, though I do not recall if this is used in the current implementation.  Generally, analyzers often have O(M*N) because we are joining two unindexed collections together.  I have not yet seen any user pain related to the slow performance of individual analyzers.